### DISCUSSION TITLE
Confusing cols() and linguistics for coders

### DISCUSSION THREAD

**Discussion by emcastro** [dflib/dflib#423] [General]
```Java
%maven org.dflib:dflib-jupyter:1.1.0
```

# Confusing cols()

Most of usages of `cols()` use it to specify destination columns:

```Java

DataFrame df = DataFrame.foldByRow("name", "phones").of(
        "Cosin", List.of("111-555-5555","111-666-6666","111-777-7777"),
        "O'Hara", List.of("222-555-5555"));

df.cols("phones_count").merge($col("phones",List.class).mapVal(l -> l.size())); // new column
df.cols("name").merge($str("name").mapVal(n->n.toUpperCase())); // replace existing column
```

    
    name   phones                                    
    ------ ------------------------------------------
    COSIN  [111-555-5555, 111-666-6666, 111-777-7777]
    O'HARA [222-555-5555]                            
    2 rows x 2 columns

However, sometimes `cols()` indicates the sources columns:

```Java
df.cols("name","non-existent").as("x","y")
```

    
    x      phones                                     y   
    ------ ------------------------------------------ ----
    Cosin  [111-555-5555, 111-666-6666, 111-777-7777] null
    O'Hara [222-555-5555]                             null
    2 rows x 3 columns

# Linguistics of Dataframes

In my humble opinion, having diverse semantics for `cols` makes the API difficult to learn. Furthermore, I find it challenging to locate the appropriate method when browsing through the IDE's completion menu (Ctrl+Tab). Should I look for what I need at the DataFrame level, under `cols`, or under `rows`? Should source columns go into `cols`, or target columns?

This is fundamentally a linguistic issue. For speakers of classical Latin, who place the verb at the end of sentences, DFLib's syntax might feel natural. However, for speakers of French, English, or Chinese – languages where the verb typically follows the subject (e.g., DataFrame) – thinking about most of the complements (such as `cols` and `rows`) before specifying the action requires extra effort.

# Facade

As for now, I use DFLib through a facade class that does that does not attempt to solve all the problems – raw DFLib is there available for that. The facade only encapsulates the most useful patterns of DFLib code:
* addColumn
* addExpanded
* expandRows
* expandRowsWithIndex
* selectRows
* groupBy
* moveColumnBefore
* moveColumnAfter
* renameColumn
* replaceColumn
* keepColums
* ... and any pattern that I would find usefull.
  
All the methods are implemented with a few lines of DFLib. They reuse the standards Exp, RowToValueMapper.

It something is missing, programmers are free to fallback to the standard DFLib.

I consider that the cost of reshaping columns is marginal – maybe I'm wrong. So, if I want to replace the content of a columns:
```java
df.cols("mycol").merge($col("mycol").mapVal(v->someaction(v)))`)
```
I do it simply in two steps:
```java
var fdf = ...; // fdf: Facade DF
fdf = fdf.addColumn("my_tranformed_col", $col("mycol").mapVal(v->someaction(v)));
fdf = fdf.replaceColumn("mycol","my_tranformed_col"); // if I inadvertently  mix source and destination, the method raises an exception.
```
In the later, most of programmers will have an idea of what my code does, even if they don't know the API.

# Conclusion
Maybe, DFLib should come in a two-layered API. The core API that does the job in the most efficient way (and the current DFLib seems to be very successful there), and an higher-level API whose job is to be easy to read/code with, even if it does not do everyting.

**Comment by andrus** [comment]
We actually had very similar "facade" methods in the early versions of DataFrame API, but eventually switched everything to `cols()` / `rows()`. I'd really hate to go back. The original API became a mess of semantically inconsistent and functionally limited methods. Reminded me of my struggles with pandas. Practically all the methods that that you mention have a history of internal struggle to make them work for a wider set of scenarios. 

> Most of usages of cols() use it to specify destination columns:
> However, sometimes cols() indicates the sources columns:

They _always_ indicate target columns. `as` method is just an extra renaming step. When we switched to the new API, it took me a bit of time to adjust, but once I internalized that rename / replace / add are all columnset operations, things started to look natural. 

And of course, `cols()` / `rows()` can and should be improved. Maybe some of the "linguistics" can be baked into them? We already discussed how the expansion operation should behave. The next improvement may be about moving columns (and rows) around in a more transparent fashion that doe not require to specify the entire total order.

**Reply by emcastro** [reply]
I agree that the API using `cols` and `rows` provide the flexibility to handle all scenarios. The facade should not cover every possible cases, as this would end up in the mess that you encountered.
I'll give you feedback after testing it with my dev team.

Maybe the problem will be resolved when the documentation is completed (a very big task).
> They always indicate target columns. as method is just an extra renaming step

This is a point that I hadn't understood until now.

**Comment by andrus** [comment]
> The facade should not cover every possible cases, as this would end up in the mess that you encountered. I'll give you feedback after testing it with my dev team.

Sounds good. BTW, you may look at `DataFrame.map(..)`. It allows to implement custom transformations without a wrapper, and preserve the ability to chain standard and custom operations at the cost of some extra verbosity:

```java
var df1 = df.map(d -> Util.replaceColumn(d, "mycol","my_tranformed_col"));
```

> This is a point that I hadn't understood until now.

Just realized we never say it explicitly in the docs. I will add a note.

**Comment by andrus** [comment]
Just added a new docs chapter: https://dflib.org/dflib/docs/1.x/#understanding_columnset
